File: lib/engine/matching.ts
// ============================================================================
// Aegis Vault — Order Matching Engine
// Implements price-time priority matching with partial fills.
// ============================================================================

import {
  type Market,
  type Order,
  type Fill,
  type Position,
  type TraderAccount,
  Side,
  OrderType,
  BPS_DENOMINATOR,
} from './types';

/**
 * Calculate the fee for a fill.
 * fee = fillSize * fillPrice * feeBps / 10000
 */
export function calculateFee(
  fillSize: number,
  fillPrice: number,
  feeBps: number
): number {
  return (fillSize * fillPrice * feeBps) / BPS_DENOMINATOR;
}

/**
 * Insert an order into the book at the correct position (price-time priority).
 * Bids: descending by price, then ascending by timestamp.
 * Asks: ascending by price, then ascending by timestamp.
 */
export function insertOrderIntoBook(market: Market, order: Order): void {
  const book = order.side === Side.Long ? market.bids : market.asks;

  let insertIdx = book.length;
  for (let i = 0; i < book.length; i++) {
    if (order.side === Side.Long) {
      // Bids: higher price first; ties broken by earlier timestamp
      if (
        order.price > book[i].price ||
        (order.price === book[i].price && order.timestamp < book[i].timestamp)
      ) {
        insertIdx = i;
        break;
      }
    } else {
      // Asks: lower price first; ties broken by earlier timestamp
      if (
        order.price < book[i].price ||
        (order.price === book[i].price && order.timestamp < book[i].timestamp)
      ) {
        insertIdx = i;
        break;
      }
    }
  }

  book.splice(insertIdx, 0, order);
  updateBestPrices(market);
}

/**
 * Update the best bid/ask on a market from the current book.
 */
export function updateBestPrices(market: Market): void {
  market.bestBid = market.bids.length > 0 ? market.bids[0].price : 0;
  market.bestAsk = market.asks.length > 0 ? market.asks[0].price : Infinity;
}

/**
 * Match an incoming taker order against the book.
 * Returns a list of fills. If the order is partially filled and is a limit,
 * the remainder is inserted into the book as a resting order.
 * Market orders that can't be fully filled throw an error.
 */
export function matchOrder(market: Market, takerOrder: Order): Fill[] {
  const fills: Fill[] = [];
  const oppositeBook =
    takerOrder.side === Side.Long ? market.asks : market.bids;

  while (takerOrder.size > 0 && oppositeBook.length > 0) {
    const bestResting = oppositeBook[0];

    // Price check: for limit orders, ensure price is acceptable
    if (takerOrder.orderType === OrderType.Limit) {
      if (
        takerOrder.side === Side.Long &&
        takerOrder.price < bestResting.price
      ) {
        break; // Taker buy price too low
      }
      if (
        takerOrder.side === Side.Short &&
        takerOrder.price > bestResting.price
      ) {
        break; // Taker sell price too high
      }
    }

    // Determine fill size and price
    const fillSize = Math.min(takerOrder.size, bestResting.size);
    const fillPrice = bestResting.price; // Maker price wins

    const fee = calculateFee(fillSize, fillPrice, market.feeBps);

    fills.push({
      maker: bestResting.trader,
      taker: takerOrder.trader,
      price: fillPrice,
      size: fillSize,
      fee,
      side: takerOrder.side,
    });

    // Update sizes
    takerOrder.size -= fillSize;
    bestResting.size -= fillSize;

    // Update open interest
    market.openInterest += fillSize;

    // Remove fully filled resting order
    if (bestResting.size <= 0) {
      oppositeBook.shift();
    }
  }

  // If market order wasn't fully filled, throw error
  if (takerOrder.orderType === OrderType.Market && takerOrder.size > 0) {
    // Revert fills (in a real system we'd be transactional)
    throw new Error('InsufficientLiquidity');
  }

  // If limit order has remaining size, insert into book
  if (takerOrder.orderType === OrderType.Limit && takerOrder.size > 0) {
    insertOrderIntoBook(market, takerOrder);
  }

  updateBestPrices(market);
  return fills;
}

/**
 * Remove an order from the book by ID.
 * Returns the removed order or null if not found.
 */
export function cancelOrder(
  market: Market,
  orderId: number,
  side: Side
): Order | null {
  const book = side === Side.Long ? market.bids : market.asks;
  const idx = book.findIndex((o) => o.id === orderId);
  if (idx === -1) return null;

  const [removed] = book.splice(idx, 1);
  updateBestPrices(market);
  return removed;
}

/**
 * Apply a fill to open/update positions for both maker and taker.
 */
export function applyFillToPositions(
  fill: Fill,
  makerAccount: TraderAccount,
  takerAccount: TraderAccount
): void {
  const makerSide = fill.side === Side.Long ? Side.Short : Side.Long;

  updatePosition(takerAccount, fill.side, fill.price, fill.size, 'market-1');
  updatePosition(makerAccount, makerSide, fill.price, fill.size, 'market-1');
}

/**
 * Update or create a position for a trader.
 */
function updatePosition(
  account: TraderAccount,
  side: Side,
  price: number,
  size: number,
  marketAddr: string
): void {
  const existing = account.positions.find(
    (p) => p.market === marketAddr && p.side === side
  );

  if (existing) {
    // Average in: weighted average entry price
    const totalSize = existing.size + size;
    existing.entryPrice =
      (existing.entryPrice * existing.size + price * size) / totalSize;
    existing.size = totalSize;
  } else {
    account.positions.push({
      market: marketAddr,
      side,
      size,
      entryPrice: price,
      marginAllocated: 0,
      unrealizedPnl: 0,
    });
  }
}

/**
 * Create a default empty market.
 */
export function createMarket(
  address: string,
  baseAsset: string,
  quoteAsset: string,
  feeBps: number = 10,
  insuranceFundCutBps: number = 8000
): Market {
  return {
    address,
    baseAsset,
    quoteAsset,
    bids: [],
    asks: [],
    bestBid: 0,
    bestAsk: Infinity,
    openInterest: 0,
    markPrice: 0,
    indexPrice: 0,
    feeBps,
    insuranceFundCutBps,
  };
}

File: engine/src/matching.rs
use crate::types::{Fill, Market, Order, OrderType, Side, TraderAccount, BPS_DENOMINATOR};

pub fn calculate_fee(fill_size: f64, fill_price: f64, fee_bps: u32) -> f64 {
    (fill_size * fill_price * (fee_bps as f64)) / BPS_DENOMINATOR
}

pub fn insert_order_into_book(market: &mut Market, order: Order) {
    let book = match order.side {
        Side::Long => &mut market.bids,
        Side::Short => &mut market.asks,
    };

    let mut insert_idx = book.len();
    for i in 0..book.len() {
        if order.side == Side::Long {
            if order.price > book[i].price || (order.price == book[i].price && order.timestamp < book[i].timestamp) {
                insert_idx = i;
                break;
            }
        } else {
            if order.price < book[i].price || (order.price == book[i].price && order.timestamp < book[i].timestamp) {
                insert_idx = i;
                break;
            }
        }
    }

    book.insert(insert_idx, order);
    update_best_prices(market);
}

pub fn update_best_prices(market: &mut Market) {
    market.best_bid = if market.bids.is_empty() { 0.0 } else { market.bids[0].price };
    market.best_ask = if market.asks.is_empty() { f64::INFINITY } else { market.asks[0].price };
}

pub fn match_order(market: &mut Market, mut taker_order: Order) -> Result<Vec<Fill>, &'static str> {
    let mut fills = Vec::new();

    while taker_order.size > 0.0 {
        let opposite_book = match taker_order.side {
            Side::Long => &mut market.asks,
            Side::Short => &mut market.bids,
        };

        if opposite_book.is_empty() {
            break;
        }

        let best_resting_price = opposite_book[0].price;

        if taker_order.order_type == OrderType::Limit {
            if taker_order.side == Side::Long && taker_order.price < best_resting_price {
                break;
            }
            if taker_order.side == Side::Short && taker_order.price > best_resting_price {
                break;
            }
        }

        let fill_size = taker_order.size.min(opposite_book[0].size);
        let fill_price = best_resting_price;
        let fee = calculate_fee(fill_size, fill_price, market.fee_bps);

        fills.push(Fill {
            maker: opposite_book[0].trader.clone(),
            taker: taker_order.trader.clone(),
            price: fill_price,
            size: fill_size,
            fee,
            side: taker_order.side,
        });

        taker_order.size -= fill_size;
        
        let opposite_book = match taker_order.side {
            Side::Long => &mut market.asks,
            Side::Short => &mut market.bids,
        };
        opposite_book[0].size -= fill_size;

        market.open_interest += fill_size;

        if opposite_book[0].size <= 0.0 {
            opposite_book.remove(0);
        }
    }

    if taker_order.order_type == OrderType::Market && taker_order.size > 0.0 {
        return Err("InsufficientLiquidity");
    }

    if taker_order.order_type == OrderType::Limit && taker_order.size > 0.0 {
        insert_order_into_book(market, taker_order);
    } else {
        update_best_prices(market);
    }

    Ok(fills)
}

pub fn cancel_order(market: &mut Market, order_id: u64, side: Side) -> Option<Order> {
    let book = match side {
        Side::Long => &mut market.bids,
        Side::Short => &mut market.asks,
    };

    if let Some(idx) = book.iter().position(|o| o.id == order_id) {
        let removed = book.remove(idx);
        update_best_prices(market);
        Some(removed)
    } else {
        None
    }
}

pub fn apply_fill_to_positions(
    fill: &Fill,
    maker_account: &mut TraderAccount,
    taker_account: &mut TraderAccount,
) {
    let maker_side = if fill.side == Side::Long { Side::Short } else { Side::Long };

    update_position(taker_account, fill.side, fill.price, fill.size, "market-1");
    update_position(maker_account, maker_side, fill.price, fill.size, "market-1");
}

fn update_position(
    account: &mut TraderAccount,
    side: Side,
    price: f64,
    size: f64,
    market_addr: &str,
) {
    if let Some(existing) = account.positions.iter_mut().find(|p| p.market == market_addr && p.side == side) {
        let total_size = existing.size + size;
        existing.entry_price = (existing.entry_price * existing.size + price * size) / total_size;
        existing.size = total_size;
    } else {
        account.positions.push(crate::types::Position {
            market: market_addr.to_string(),
            side,
            size,
            entry_price: price,
            margin_allocated: 0.0,
            unrealized_pnl: 0.0,
        });
    }
}

pub fn create_market(
    address: &str,
    base_asset: &str,
    quote_asset: &str,
    fee_bps: Option<u32>,
    insurance_fund_cut_bps: Option<u32>,
) -> Market {
    Market {
        address: address.to_string(),
        base_asset: base_asset.to_string(),
        quote_asset: quote_asset.to_string(),
        bids: Vec::new(),
        asks: Vec::new(),
        best_bid: 0.0,
        best_ask: f64::INFINITY,
        open_interest: 0.0,
        mark_price: 0.0,
        index_price: 0.0,
        fee_bps: fee_bps.unwrap_or(10),
        insurance_fund_cut_bps: insurance_fund_cut_bps.unwrap_or(8000),
    }
}
